### PR TITLE
Adds universal mocks for testing and JDate::getInstance.

### PR DESCRIPTION
**Pull request by eddieajau**
Added support for getting universal mock objects in JoomlaTestCase providing getters for JApplication, JConfig, JDatabase and JSession mocks.

Converted JDatabaseQuery test to use the universal database mock.

Improved setup and teardown of rules field test in JForm.

Added tests for JDate (tests involving language are incomplete because of tight coupling to the CMS).

Added static getInstance method to JDate as an alias for new JDate.

### CODE REVIEW

File: tests/includes/JoomlaTestCase.php [hunk 1]
```diff
@@ -214,4 +330,60 @@ abstract class JoomlaTestCase extends PHPUnit_Framework_TestCase
...
+
+		return JApplicationGlobalMock::create($this);
+	}
+
+	/**
+	 * Gets a mock database object.
+	 *
+	 * @return  object
+	 *
+	 * @return  11.3
```
Comment (ianmacl) [inline tests/includes/JoomlaTestCase.php]: two returns?

File: tests/suite/joomla/database/JDatabaseMock.php [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+	 * @return  object
+	 *
+	 * @since   11.3
+	 */
+	public static function create($test)
+	{
+		// Collect all the relevant methods in JDatabase (exlcude most ADODB compatibility methods).
+		$methods = array(
+			'addQuoted',
+// 			'BeginTrans',
```
Comment (ianmacl) [inline tests/suite/joomla/database/JDatabaseMock.php]: do we need the commented code in there?

File: tests/suite/joomla/session/JSessionMock.php [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+/**
+ * Mock class for JSession.
+ *
+ * @package  Joomla.UnitTest
+ * @since    11.3
+ */
+class JSessionGlobalMock
+{
+	/**
+	 * Creates and instance of the mock JDatabase object.
```
Comment (ianmacl) [inline tests/suite/joomla/session/JSessionMock.php]: minor
s/and/an/

File: tests/suite/joomla/utilities/JDateTest.php [hunk 4]
```diff
@@ -0,0 +1,1050 @@
...
+				true,
+				'122007 114456',
+			),
+			'mmddyyGMT' => array(
+				'mdy His',
+				false,
+				'122007 164456',
+			),
+			// TODO  Need to fix up language string/file dependancies before we can do these properly.
+// 			'Long' => array(
```
Comment (ianmacl) [inline tests/suite/joomla/utilities/JDateTest.php:155]: do we need all this commented out code?

Comment (eddieajau) [inline tests/suite/joomla/utilities/JDateTest.php:155]: Yes, because the platform isn't loading language files yet. Once it does, we can add those tests.

**Comment by eddieajau** [comment]
Fixed up cosmetic issues.

**Comment by eddieajau** [comment]
Added a bit more to the branch: getSetUpOperation and getTearDownOperation to support InnoDB tables with foreign key constraints (had to fix the JAccess tests - parent::setUp was not being run properly for all tests).

**Comment by eddieajau** [comment]
A few more bug fixes, plus ported JTableNested tests from the CMS.

**Comment by ianmacl** [comment]
# Test Results

Total Tests: 1280
Assertions: 2217
Failures: 0
Errors: 0
Test Time: 56.449255

File: tests/includes/JoomlaDatabaseTestCase.php [hunk 5]
```diff
@@ -88,6 +201,15 @@ abstract class JoomlaDatabaseTestCase extends PHPUnit_Extensions_Database_TestCa
...
 	}
 
+	/**
+	 * Sets up the fixture.
+	 *
+	 * This method is called before a test is executed.
+	 *
+	 * @return  void
+	 *
+     * @since   11.1
```
Comment (ianmacl) [inline tests/includes/JoomlaDatabaseTestCase.php]: is this an indent problem?

Comment (eddieajau) [inline tests/includes/JoomlaDatabaseTestCase.php]: Fixed.

**Comment by ianmacl** [comment]
# Test Results

Total Tests: 1280
Assertions: 2217
Failures: 1
Errors: 0
Test Time: 56.008952

**Comment by ianmacl** [comment]
Test failure is result of time not being mocked out (sometimes the expected result is off by a second).

**Pull request merged**